blog/2021-03-16-BatchSolvingWithAmq.adoc

```diff
@@ -0,0 +1,182 @@
+            }
+        });
+    }
+    ...
+----
+<1> Inject `ObjectMapper` to unmarshall the JSON message payload.
+<2> `Emitter` sends response messages to the `solver_response` channel.
+<3> Inject a `SolverFactory` and build a `Solver`.
+<4> The `@Incoming` annotation makes the method listen for incoming messages from the `solver_request` channel.
+<5> By accepting `Message` as a parameter, you have full control over acknowledgement of the message.
+The generic type of the `Message` is `String`, because the message contains the `SolverRequest` serialized to a JSON String.
+Finally, the return type `+CompletionStage<Void>+` enables an asynchronous acknowledgement.
+See https://smallrye.io/smallrye-reactive-messaging/smallrye-reactive-messaging/2.8/model/model.html#consuming-messages[Consuming Messages] for more details.
+<6> Return a `+CompletionStage<Void>+` to satisfy the method contract and avoid blocking the thread.
+<7> Unmarshall the JSON payload. If it's not possible, reject the message.
+<8> Solve the input timetabling problem and then send a reply (see the next figure).
+<9> In case any exception occurs, include information about the exception into the response.
+
+
+The example below shows how to reply and acknowledge the original request message:
+[source,java]
+----
+private void replySuccess(Message<String> solverRequestMessage) {
+...
+solverResponseEmitter.send(jsonResponse)
+    .thenAccept(x -> solverRequestMessage.ack()); // <1>
+...
+}
+----
+<1> `thenAccept()` defines what happens when the AMQ broker acknowledges the response message sent via the `Emitter`.
+In this case, the request message is acknowledged.
+This way, the request message is never lost even if the Solver Server dies.
+
+To understand how the channels correspond to messaging queues, see the `application.properties` file located in `src/main/resources`:
+[source]
+----
+# Configure the AMQ source
+mp.messaging.incoming.solver_request.connector=smallrye-amqp # <1>
+mp.messaging.incoming.solver_request.durable=true # <2>
+mp.messaging.incoming.solver_request.failure-strategy=reject # <3>
+
+# Configure the AMQ sink
+mp.messaging.outgoing.solver_response.connector=smallrye-amqp
+mp.messaging.outgoing.solver_response.durable=true
+----
+<1> Use the `smallrye-amqp` connector for the `solver_request` channel.
+<2> To have the AMQ broker persist messages, make the queue durable.
+<3> If a message is rejected, the broker redirects it to a dead letter queue and proceeds with the next message.
+
+Every property contains the channel name. By default, it matches the name of the queue at the AMQ broker.
+
+== Running the quickstart
+*Prerequisites:* install https://docs.docker.com/get-docker/[docker] and https://docs.docker.com/compose/install[docker-compose].
+
+. git clone https://github.com/kiegroup/optaplanner-quickstarts
+. cd optaplanner-quickstarts/amq-quarkus-school-timetabling
+. ./run.sh
+. Open http://localhost:8080 in a browser and click the Solve button.
```

Comment:
Can we have the first or last usage of the word OptaPlanner link to https://www.optaplanner.org ? That helps for link building for those that consume our atom feed and republish the article.